nixos/modules/services/web-apps/mastodon.nix

```diff
@@ -700,28 +714,157 @@ in {
+            ${nginxCommonHeaders}
+          '';
+        };
+
+        locations."^~ /packs/" = {
+          tryFiles = "$uri =404";
+          priority = 2250;
+
+          extraConfig = ''
+            add_header Cache-Control 'public, max-age=2419200, must-revalidate';
+            ${nginxCommonHeaders}
+          '';
+        };
+
+        locations."^~ /shortcuts/" = {
+          tryFiles = "$uri =404";
+          priority = 2260;
+
+          extraConfig = ''
+            add_header Cache-Control 'public, max-age=2419200, must-revalidate';
+            ${nginxCommonHeaders}
+          '';
+        };
+
+        locations."^~ /sounds/" = {
+          tryFiles = "$uri =404";
+          priority = 2270;
+
+          extraConfig = ''
+            add_header Cache-Control 'public, max-age=2419200, must-revalidate';
+            ${nginxCommonHeaders}
+          '';
+        };
+
+        locations."^~ /system/" = {
+          tryFiles = "$uri =404";
+          alias = "/var/lib/mastodon/public-system/";
+          priority = 2280;
+
+          extraConfig = ''
+            add_header Cache-Control 'public, max-age=2419200, immutable';
+            ${nginxCommonHeaders}
+          '';
+        };
+
+        locations."^~ /api/v1/streaming/" = {
+          proxyPass = if cfg.enableUnixSocket then "http://unix:/run/mastodon-streaming/streaming.socket" else "http://127.0.0.1:${toString cfg.streamingPort}/";
           proxyWebsockets = true;
+          priority = 2300;
+
+          extraConfig = ''
+            ${nginxProxyHeaders}
+            proxy_buffering off;
+            proxy_redirect off;
+
+            tcp_nodelay on;
+          '';
         };
 
-        locations."/api/v1/streaming/" = {
-          proxyPass = (if cfg.enableUnixSocket then "http://unix:/run/mastodon-streaming/streaming.socket" else "http://127.0.0.1:${toString(cfg.streamingPort)}/");
+        locations."@proxy" = {
+          proxyPass = if cfg.enableUnixSocket then "http://unix:/run/mastodon-web/web.socket" else "http://127.0.0.1:${toString cfg.webPort}";
           proxyWebsockets = true;
+          priority = 2400;
+
+          extraConfig = ''
+            ${nginxProxyHeaders}
+            proxy_pass_header Server;
+
+            proxy_buffering on;
+            proxy_redirect off;
+
+            proxy_cache all_cache;
+            proxy_cache_valid 200 7d;
+            proxy_cache_valid 410 24h;
+            proxy_cache_use_stale error timeout updating http_500 http_502 http_503 http_504;
+            add_header X-Cached $upstream_cache_status;
+            ${lib.optionalString config.services.nginx.virtualHosts.${cfg.localDomain}.http3 ''
+              add_header Alt-Svc 'h3=":$server_port"; ma=86400';
+            ''}
+
+            tcp_nodelay on;
+          '';
         };
+
+        extraConfig = ''
+          add_header Strict-Transport-Security 'max-age=63072000; includeSubDomains';
```

Comment:
I don't see this used outside a location directive at upstreams recommendation?

Reply:
See - https://github.com/mastodon/mastodon/pull/19018#issuecomment-1301963161

Reply:
Wasn't it the point of this exact PR you linked, that this header outside location directives should be removed? https://github.com/mastodon/mastodon/pull/19018/files

Please let's just follow current upstream for now. Just for the sake of maintainability.

Reply:
Update PR in upstream - https://github.com/mastodon/mastodon/pull/19644